server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/aws/AwsS3ObjectStorage.java

```diff
@@ -19,27 +19,105 @@
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.services.s3.transfer.TransferManagerBuilder;
 import com.google.common.collect.ImmutableSet;
 import com.google.inject.Module;
 
 public class AwsS3ObjectStorage {
 
     private static final Iterable<Module> JCLOUDS_MODULES =
         ImmutableSet.of(new SLF4JLoggingModule());
+    public static final int MAX_UPLOAD_THREADS = 5;
+    public static Size MULTIPART_UPLOAD_THRESHOLD;
+
+    static {
+        try {
+            MULTIPART_UPLOAD_THRESHOLD = Size.parse("5M");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
 
     public static ObjectStorageBlobsDAOBuilder.RequireContainerName daoBuilder(AwsS3AuthConfiguration configuration) {
         return ObjectStorageBlobsDAOBuilder.forBlobStore(new BlobStoreBuilder(configuration));
     }
 
+    public static Optional<PutBlobFunction> putBlob(BlobId.Factory blobIdFactory, ContainerName containerName, AwsS3AuthConfiguration configuration) {
+        return Optional.of((blob) -> {
+            File file = null;
+            try {
+                file = File.createTempFile(UUID.randomUUID().toString(), ".tmp");
+                FileUtils.copyToFile(blob.getPayload().openStream(), file);
+
+                return put(blobIdFactory, containerName, configuration, blob, file);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            } finally {
+                if (file != null) {
+                    FileUtils.deleteQuietly(file);
+                }
+            }
+        });
+    }
+
+    private static BlobId put(BlobId.Factory blobIdFactory, ContainerName containerName, AwsS3AuthConfiguration configuration, Blob blob, File file) {
+        try {
+            PutObjectRequest request = new PutObjectRequest(containerName.value(),
+                blob.getMetadata().getName(),
+                file);
+
+            return blobIdFactory.from(getTransferManager(configuration)
+                .upload(request)
+                .waitForUploadResult()
+                .getETag());
+        } catch (AmazonClientException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static TransferManager getTransferManager(AwsS3AuthConfiguration configuration) {
+        AmazonS3 amazonS3 = AmazonS3ClientBuilder
+            .standard()
+            .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials(configuration.getAccessKeyId(), configuration.getSecretKey())))
+            .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(configuration.getEndpoint(), null))
+            .build();
+
+        return TransferManagerBuilder
+            .standard()
+            .withS3Client(amazonS3)
+            .withMultipartUploadThreshold(MULTIPART_UPLOAD_THRESHOLD.getValue())
+            .withExecutorFactory(() -> Executors.newFixedThreadPool(MAX_UPLOAD_THREADS))
```

Comment:
This is not a good idea: we don't want to create an executor everytime we send a file. Please use single threadpool for all uploads.
And please give it a name, otherwise, it's a hell to monitor thread usage.